File: src/camera.cpp
#include "camera.hpp"

#include <math.h>
#include <assert.h>

#include "attitude-utils.hpp"

namespace lost {

/**
 * Converts from a 3D point in space to a 2D point on the camera sensor.
 * Assumes that X is the depth direction and that it points away from the center of the sensor, i.e., any vector (x, 0, 0) will be at (xResolution/2, yResolution/2) on the sensor.
 */
Vec2 Camera::SpatialToCamera(const Vec3 &vector) const {
    // can't handle things behind the camera.
    assert(vector.x > 0);
    // TODO: is there any sort of accuracy problem when vector.y and vector.z are small?

    decimal focalFactor = focalLength/vector.x;

    decimal yPixel = vector.y*focalFactor;
    decimal zPixel = vector.z*focalFactor;

    return { -yPixel + xCenter, -zPixel + yCenter };
}

/**
 * Gives a point in 3d space that could correspond to the given vector, using the same
 * coordinate system described for SpatialToCamera.
 * Not all vectors returned by this function will necessarily have the same magnitude.
 * @return A vector in 3d space corresponding to the given vector, with x-component equal to 1
 * @warning Other functions rely on the fact that returned vectors are placed one unit away (x-component equal to 1). Don't change this behavior!
 */
Vec3 Camera::CameraToSpatial(const Vec2 &vector) const {
    assert(InSensor(vector));

    // isn't it interesting: To convert from center-based to left-corner-based coordinates is the
    // same formula; f(x)=f^{-1}(x) !
    decimal xPixel = -vector.x + xCenter;
    decimal yPixel = -vector.y + yCenter;

    return {
        1,
        xPixel / focalLength,
        yPixel / focalLength,
    };
}

/// Returns whether a given pixel is actually in the camera's field of view
bool Camera::InSensor(const Vec2 &vector) const {
    // if vector.x == xResolution, then it is at the leftmost point of the pixel that's "hanging
    // off" the edge of the image, so vector is still in the image.
    return vector.x >= 0 && vector.x <= xResolution
        && vector.y >= 0 && vector.y <= yResolution;
}

decimal FovToFocalLength(decimal xFov, decimal xResolution) {
    return xResolution / 2.0f / tan(xFov/2);
}

decimal FocalLengthToFov(decimal focalLength, decimal xResolution, decimal pixelSize) {
    return atan(xResolution/2 * pixelSize / focalLength) * 2;
}

decimal Camera::Fov() const {
    return FocalLengthToFov(focalLength, xResolution, 1.0);
}

}

File: python/camera.py
"""
====================================================
CAMERA MODEL
====================================================

Purpose:
    Describes the camera's optical properties and provides
    conversions between pixel coordinates and spatial vectors.

    The camera coordinate system has:
        - X pointing away from the sensor (boresight direction)
        - Y and Z spanning the sensor plane
        - Origin at the principal point (center of sensor)

Mathematical Background:
    The pinhole camera model maps a 3D point (x, y, z) in camera
    coordinates to a 2D pixel (u, v) via:
        u = -y * f / x + cx
        v = -z * f / x + cy
    where f is focal length in pixels, (cx, cy) is the principal point.

    The inverse mapping places points on a plane at x=1:
        y' = -(u - cx) / f
        z' = -(v - cy) / f
        returns (1, y', z')

Reference:
    Original C++ source: src/camera.hpp, src/camera.cpp
"""

from __future__ import annotations

import math

from math_utils import Vec2, Vec3, angle_unit


class Camera:
    """Full description of a camera.

    Stores focal length, principal point, and resolution.
    Provides spatial↔pixel coordinate conversions.

    Args:
        focal_length: Focal length in pixels
        x_center: Principal point X coordinate (pixels)
        y_center: Principal point Y coordinate (pixels)
        x_resolution: Sensor width in pixels
        y_resolution: Sensor height in pixels

    Reference:
        Original C++ source: src/camera.hpp:8-58, src/camera.cpp:1-70
    """

    def __init__(self, focal_length: float,
                 x_center: float, y_center: float,
                 x_resolution: int, y_resolution: int):
        self._focal_length = float(focal_length)
        self._x_center = float(x_center)
        self._y_center = float(y_center)
        self._x_resolution = int(x_resolution)
        self._y_resolution = int(y_resolution)

    @classmethod
    def from_center(cls, focal_length: float,
                    x_resolution: int, y_resolution: int) -> Camera:
        """Create a camera with principal point at the sensor center."""
        return cls(focal_length,
                   x_resolution / 2.0, y_resolution / 2.0,
                   x_resolution, y_resolution)

    def spatial_to_camera(self, vector: Vec3) -> Vec2:
        """Convert a 3D spatial point to 2D pixel coordinates.

        Assumes X is the depth direction pointing away from the sensor.
        Any vector (x, 0, 0) maps to the principal point.

        Reference:
            Original C++ source: src/camera.cpp:15-26
        """
        assert vector.x > 0, f"Point behind camera: x={vector.x}"

        focal_factor = self._focal_length / vector.x

        y_pixel = vector.y * focal_factor
        z_pixel = vector.z * focal_factor

        return Vec2(-y_pixel + self._x_center, -z_pixel + self._y_center)

    def camera_to_spatial(self, vector: Vec2) -> Vec3:
        """Convert 2D pixel coordinates to a 3D spatial direction.

        Returns a vector with x-component equal to 1 (placed one unit away
        along the boresight). The returned vector is NOT necessarily
        a unit vector — call .normalize() if you need one.

        WARNING: Other functions rely on x-component being 1.

        Reference:
            Original C++ source: src/camera.cpp:35-48
        """
        x_pixel = -vector.x + self._x_center
        y_pixel = -vector.y + self._y_center

        return Vec3(
            1,
            x_pixel / self._focal_length,
            y_pixel / self._focal_length,
        )

    def in_sensor(self, vector: Vec2) -> bool:
        """Check whether a pixel coordinate is within the sensor bounds.

        Returns True if the point is on or within the sensor edges.

        Reference:
            Original C++ source: src/camera.cpp:51-56
        """
        return (0 <= vector.x <= self._x_resolution
                and 0 <= vector.y <= self._y_resolution)

    @property
    def x_resolution(self) -> int:
        return self._x_resolution

    @property
    def y_resolution(self) -> int:
        return self._y_resolution

    @property
    def focal_length(self) -> float:
        return self._focal_length

    @property
    def fov(self) -> float:
        """Horizontal field of view in radians.

        Reference:
            Original C++ source: src/camera.cpp:66-68
        """
        return focal_length_to_fov(self._focal_length, self._x_resolution, 1.0)

    @focal_length.setter
    def focal_length(self, value: float) -> None:
        self._focal_length = float(value)


# ─────────────────────────────────────────
# FOV <-> FOCAL LENGTH CONVERSIONS
# ─────────────────────────────────────────


def fov_to_focal_length(x_fov: float, x_resolution: int) -> float:
    """Convert horizontal field of view to focal length in pixels.

    Reference:
        Original C++ source: src/camera.cpp:58-60
    """
    return x_resolution / 2.0 / math.tan(x_fov / 2)


def focal_length_to_fov(focal_length: float, x_resolution: int,
                        pixel_size: float) -> float:
    """Convert focal length (with pixel size) to horizontal field of view.

    Reference:
        Original C++ source: src/camera.cpp:62-64
    """
    return math.atan2(x_resolution / 2 * pixel_size, focal_length) * 2
